nuxeo-runtime/nuxeo-stream/src/main/java/org/nuxeo/lib/stream/log/kafka/KafkaLogTailer.java

```diff
@@ -0,0 +1,408 @@
+import org.apache.commons.logging.LogFactory;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.InterruptException;
+import org.apache.kafka.common.errors.WakeupException;
+import org.apache.kafka.common.utils.Bytes;
+import org.nuxeo.lib.stream.log.LogOffset;
+import org.nuxeo.lib.stream.log.LogPartition;
+import org.nuxeo.lib.stream.log.LogRecord;
+import org.nuxeo.lib.stream.log.LogTailer;
+import org.nuxeo.lib.stream.log.RebalanceException;
+import org.nuxeo.lib.stream.log.RebalanceListener;
+import org.nuxeo.lib.stream.log.internals.LogOffsetImpl;
+
+import java.io.ByteArrayInputStream;
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectInputStream;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.ConcurrentModificationException;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Queue;
+import java.util.stream.Collectors;
+
+
+/**
+ * @since 9.2
+ */
+public class KafkaLogTailer<M extends Externalizable> implements LogTailer<M>, ConsumerRebalanceListener {
+    private static final Log log = LogFactory.getLog(KafkaLogTailer.class);
+    protected final String group;
+    protected final String prefix;
+    protected KafkaConsumer<String, Bytes> consumer;
+    protected String id;
+    protected Collection<TopicPartition> topicPartitions;
+    protected Collection<LogPartition> partitions;
+    protected final Map<TopicPartition, Long> lastOffsets = new HashMap<>();
+    protected final Map<TopicPartition, Long> lastCommittedOffsets = new HashMap<>();
+    protected final Queue<ConsumerRecord<String, Bytes>> records = new LinkedList<>();
+    // keep track of all tailers on the same namespace index even from different mq
+    protected boolean closed = false;
+    protected Collection<String> names;
+    protected RebalanceListener listener;
+    protected boolean isRebalanced = false;
+
+    protected KafkaLogTailer(String prefix, String group, Properties consumerProps) {
+        Objects.requireNonNull(group);
+        this.prefix = prefix;
+        this.group = group;
+        consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, group);
+        this.consumer = new KafkaConsumer<>(consumerProps);
+
+    }
+
+    public static <M extends Externalizable> KafkaLogTailer<M> createAndAssign(String prefix, Collection<LogPartition> partitions, String group, Properties consumerProps) {
+        KafkaLogTailer<M> ret = new KafkaLogTailer<>(prefix, group, consumerProps);
```

Comment:
Don't call `ret` your return values, call them what they are. Here: `tailer` for instance.